File: examples/java/src/main/java/io/superdurable/dex/patterns/workflow/drainchannels/internal/DrainInternalChannelsFlow.java
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.superdurable.dex.patterns.workflow.drainchannels.internal;

import com.fasterxml.jackson.core.JsonProcessingException;
import io.superdurable.dex.Attribute;
import io.superdurable.dex.Channel;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StepList;
import io.superdurable.dex.StepMovement;
import io.superdurable.dex.Wait;
import io.superdurable.dex.patterns.services.ServiceDependency;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class DrainInternalChannelsFlow implements Flow<String> {
    public static final String UPSERT_MONGO_DATA_INTERNAL_CHANNEL =
            "upsert_mongo_data_internal_channel";
    public static final String PROCESS_DATA_STATE_EXECUTION_COUNTER =
            "process_data_state_execution_counter";

    public final Attribute<Integer> processDataStateExecutionCounter =
            Attribute.define(PROCESS_DATA_STATE_EXECUTION_COUNTER, Integer.class);
    public final Channel<MongoDocument> upsertMongoData =
            Channel.define(UPSERT_MONGO_DATA_INTERNAL_CHANNEL, MongoDocument.class);

    private final ServiceDependency externalService;
    private final ServiceDependency mongoCollection;
    private final Init init = new Init();
    private final UpsertMongoRecord upsertMongoRecord = new UpsertMongoRecord();
    private final ProcessData processData = new ProcessData();
    private final Finalize finalize = new Finalize();

    public DrainInternalChannelsFlow(final ServiceDependency service) {
        this.externalService = service;
        this.mongoCollection = service;
    }

    @Override
    public StepList<String> getSteps() {
        return StepList.startStep(init)
                .otherSteps(upsertMongoRecord, processData, finalize);
    }

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(processDataStateExecutionCounter, upsertMongoData);
    }

    final class Init implements Step<String> {
        @Override
        public Class<String> getInputType() {
            return String.class;
        }

        @Override
        public StepDecision execute(final Context context, final String input) {
            processDataStateExecutionCounter.set(context, 0);
            return StepDecision.goToMulti(
                    StepMovement.of(upsertMongoRecord, null),
                    StepMovement.of(processData, input));
        }
    }

    final class UpsertMongoRecord implements Step<Void> {
        @Override
        public Class<Void> getInputType() {
            return Void.class;
        }

        @Override
        public Wait waitFor(final Context context, final Void input) {
            return Wait.until(upsertMongoData.forOne());
        }

        @Override
        public StepDecision execute(final Context context, final Void input) {
            final List<MongoDocument> documents = upsertMongoData.getConditionResults(context);
            if (documents.isEmpty()) {
                throw new IllegalStateException("No document was sent");
            }

            final MongoDocument document = documents.get(0);
            if (document == null) {
                throw new IllegalStateException("No data was sent");
            }

            try {
                mongoCollection.upsert(document);
            } catch (final JsonProcessingException e) {
                throw new RuntimeException(e);
            }

            if (document.finalCommand) {
                return StepDecision.gracefulComplete();
            }
            return StepDecision.goTo(upsertMongoRecord, null);
        }
    }

    final class ProcessData implements Step<String> {
        @Override
        public Class<String> getInputType() {
            return String.class;
        }

        @Override
        public StepDecision execute(final Context context, final String input) {
            final int executionCount = processDataStateExecutionCounter.get(context) + 1;
            processDataStateExecutionCounter.set(context, executionCount);

            final String status;
            switch (executionCount) {
                case 1:
                    status = "RECEIVED";
                    break;
                case 2:
                    status = "ACCEPTED";
                    break;
                case 3:
                    status = "PASSED";
                    break;
                default:
                    status = "ERROR";
                    break;
            }
            upsertMongoData.publish(
                    context,
                    new MongoDocument(input, status, false));

            externalService.externalApiCall(
                    "external service call to process data (e.g. notify the job seeker)");
            externalService.externalApiCall(
                    "a call to send metrics or add a log to logrepo");

            if (executionCount <= 3) {
                return StepDecision.goTo(processData, input);
            }
            return StepDecision.goTo(finalize, null);
        }
    }

    final class Finalize implements Step<Void> {
        @Override
        public Class<Void> getInputType() {
            return Void.class;
        }

        @Override
        public StepDecision execute(final Context context, final Void input) {
            upsertMongoData.publish(
                    context,
                    new MongoDocument("documentId-1", "FINALIZED", true));
            return StepDecision.gracefulComplete();
        }
    }
}

File: examples/python/dex_examples/patterns/drain-channels/internal/drain_internal_channels_flow.py
# Copyright (c) 2022-2026 Super Durable, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from dex import (
    Attribute,
    Channel,
    Context,
    Flow,
    PersistenceSchema,
    Step,
    StepDecision,
    StepList,
    StepMovement,
    Wait,
    go_to,
    go_to_multi,
    graceful_complete,
)

from dex_examples.patterns.shared.service_dependency import ServiceDependency
from dex_examples.patterns.drain_channels.internal.mongo_document import (
    MongoDocument,
)


class Finalize(Step[None]):
    def __init__(self, upsert_mongo_data: Channel[MongoDocument]) -> None:
        self.upsert_mongo_data = upsert_mongo_data

    def execute(self, context: Context, input: None) -> StepDecision:
        del input
        self.upsert_mongo_data.publish(
            context,
            MongoDocument("documentId-1", "FINALIZED", True),
        )
        return graceful_complete()


class ProcessData(Step[str]):
    def __init__(
        self,
        finalize: Finalize,
        external_service: ServiceDependency,
        upsert_mongo_data: Channel[MongoDocument],
        execution_counter: Attribute[int],
    ) -> None:
        self.finalize = finalize
        self.external_service = external_service
        self.upsert_mongo_data = upsert_mongo_data
        self.execution_counter = execution_counter

    def execute(self, context: Context, input: str) -> StepDecision:
        execution_count = self.execution_counter.get(context) + 1
        self.execution_counter.set(context, execution_count)

        statuses = {1: "RECEIVED", 2: "ACCEPTED", 3: "PASSED"}
        self.upsert_mongo_data.publish(
            context,
            MongoDocument(input, statuses.get(execution_count, "ERROR"), False),
        )

        self.external_service.external_api_call(
            "external service call to process data (e.g. notify the job seeker)"
        )
        self.external_service.external_api_call(
            "a call to send metrics or add a log to logrepo"
        )

        if execution_count <= 3:
            return go_to(ProcessData, input)
        return go_to(Finalize, None)


class UpsertMongoRecord(Step[None]):
    def __init__(
        self,
        mongo_collection: ServiceDependency,
        upsert_mongo_data: Channel[MongoDocument],
    ) -> None:
        self.mongo_collection = mongo_collection
        self.upsert_mongo_data = upsert_mongo_data

    def wait_for(self, context: Context, input: None) -> Wait:
        del context, input
        return Wait.until(self.upsert_mongo_data.for_one())

    def execute(self, context: Context, input: None) -> StepDecision:
        del input
        documents = self.upsert_mongo_data.results(context)
        if not documents:
            raise RuntimeError("No document was sent")

        document = documents[0]
        if document is None:
            raise RuntimeError("No data was sent")

        self.mongo_collection.upsert(document)

        if document.final_command:
            return graceful_complete()
        return go_to(UpsertMongoRecord, None)


class Init(Step[str]):
    def __init__(
        self,
        upsert_mongo_record: UpsertMongoRecord,
        process_data: ProcessData,
        execution_counter: Attribute[int],
    ) -> None:
        self.upsert_mongo_record = upsert_mongo_record
        self.process_data = process_data
        self.execution_counter = execution_counter

    def execute(self, context: Context, input: str) -> StepDecision:
        self.execution_counter.set(context, 0)
        return go_to_multi(
            StepMovement.of(UpsertMongoRecord, None),
            StepMovement.of(ProcessData, input),
        )


class DrainInternalChannelsFlow(Flow[str]):
    UPSERT_MONGO_DATA_INTERNAL_CHANNEL = "upsert_mongo_data_internal_channel"
    PROCESS_DATA_STATE_EXECUTION_COUNTER = "process_data_state_execution_counter"

    process_data_state_execution_counter = Attribute(
        PROCESS_DATA_STATE_EXECUTION_COUNTER,
        int,
    )
    upsert_mongo_data = Channel(UPSERT_MONGO_DATA_INTERNAL_CHANNEL, MongoDocument)

    def __init__(self, service: ServiceDependency) -> None:
        self.finalize = Finalize(self.upsert_mongo_data)
        self.process_data = ProcessData(
            self.finalize,
            service,
            self.upsert_mongo_data,
            self.process_data_state_execution_counter,
        )
        self.upsert_mongo_record = UpsertMongoRecord(service, self.upsert_mongo_data)
        self.init = Init(
            self.upsert_mongo_record,
            self.process_data,
            self.process_data_state_execution_counter,
        )

    def get_steps(self) -> StepList[str]:
        return StepList.start_step(self.init).other_steps(
            self.upsert_mongo_record,
            self.process_data,
            self.finalize,
        )

    def get_persistence_schema(self) -> PersistenceSchema:
        return PersistenceSchema.of(
            self.process_data_state_execution_counter,
            self.upsert_mongo_data,
        )
